processing/src/main/java/org/apache/druid/segment/filter/OrFilter.java

```diff
@@ -82,158 +84,151 @@ public <T> FilterBundle makeFilterBundle(
+        } else {
+          if (!bundle.hasMatcher()) {
+            indexOnlyBundles.add(bundle.getIndex());
+            merged.merge(bundle.getIndex().getIndexCapabilities());
+            // union index only bitmaps together; if all sub-filters are 'index only' bundles we will make an index only
+            // bundle ourselves, else we will use this index as a single value matcher
+            if (index == null) {
+              index = bundle.getIndex().getBitmap();
+            } else {
+              index = index.union(bundle.getIndex().getBitmap());
+            }
+          } else {
+            // index and matcher bundles must be handled separately, they will need to be a single value matcher built
+            // by doing an AND operation between the index and the value matcher
+            // (a bundle is basically an AND operation between the index and matcher if the matcher is present)
+            partialIndexBundles.add(convertBundleToMatcherOnlyBundle(bundle, bundleIndex));
           }
-          indexFilters.add(subfilter);
-          needMatcher = !index.getIndexCapabilities().isExact();
-        } else if (!allowPartialIndex) {
-          break;
         }
-      } else if (!allowPartialIndex) {
-        break;
-      }
-
-      if (needMatcher) {
-        matcherFilters.add(subfilter);
+      } else {
+        matcherOnlyBundles.add(bundle.getMatcherBundle());
       }
     }
     final long totalBitmapConstructTimeNs = System.nanoTime() - bitmapConstructionStartNs;
 
-    final ImmutableBitmap index;
-    final ImmutableBitmap partialIndex;
-    if (indexFilters.size() == filters.size()) {
-      // all or nothing
-      index = bitmapResultFactory.toImmutableBitmap(bitmapResultFactory.union(indexes));
-      partialIndex = null;
-    } else {
-      index = null;
-      if (allowPartialIndex) {
-        if (indexes.isEmpty()) {
-          partialIndex = null;
-        } else if (indexes.size() == 1) {
-          partialIndex = bitmapResultFactory.toImmutableBitmap(indexes.get(0));
-        } else {
-          partialIndex = bitmapResultFactory.toImmutableBitmap(bitmapResultFactory.union(indexes));
-        }
-      } else {
-        partialIndex = null;
-        matcherFilters.clear();
-        // need to convert to all matchers
-        matcherFilters.addAll(filters);
+
+    // if all the filters are 'index only', we can make an index only bundle
+    if (indexOnlyBundles.size() + emptyCount == filters.size()) {
+      if (index == null || index.isEmpty()) {
+        return FilterBundle.allFalse(
+            totalBitmapConstructTimeNs,
+            columnIndexSelector.getBitmapFactory().makeEmptyImmutableBitmap()
+        );
+      }
+      if (indexOnlyBundles.size() == 1) {
+        return new FilterBundle(
+            new FilterBundle.SimpleIndexBundle(
+                indexOnlyBundles.get(0).getIndexInfo(),
+                index,
+                indexOnlyBundles.get(0).getIndexCapabilities()
+            ),
```

Comment:
Why isn't this just `indexOnlyBundles.get(0)`?  In this case, the `index` had better be the same thing as the index in the index-only bundle right?

I'm not sure it really matters, just found myself wondering if there's a reason that this code needs to build a new object.

Reply:
ah, this should be just re-using the same indexBundle, i think maybe it is stale from before i reworked some stuff and i forgot to clean it up